plugin-core/src/main/java/appland/settings/AppMapSecureApplicationSettingsService.java

```diff
@@ -16,16 +20,57 @@ public final class AppMapSecureApplicationSettingsService implements AppMapSecur
         return ApplicationManager.getApplication().getService(AppMapSecureApplicationSettingsService.class);
     }
 
+    private volatile boolean isCached = false;
+    private volatile @Nullable String cachedOpenAIKey;
+
+    /**
+     * @return The stored value of the OpenAI key. Because this method is called often
+     * and is a slow operation in 2024.2+, we're caching the value to avoid too many slow calls.
+     */
     @Override
     public @Nullable String getOpenAIKey() {
-        return PasswordSafe.getInstance().getPassword(createOpenAIKey());
+        if (!isCached) {
+            try {
+                String currentKeyValue;
+                // If on EDT, load the key in a background task to avoid the warning about the slow operation
+                if (ApplicationManager.getApplication().isDispatchThread()) {
+                    var title = AppMapBundle.get("applicationSettings.openAI.loadingKey");
+
+                    // Using a task because getPassword is a slow operation in 2024.2+
+                    var task = new Task.WithResult<String, Exception>(null, title, false) {
+                        @Override
+                        protected String compute(@NotNull ProgressIndicator indicator) {
+                            return PasswordSafe.getInstance().getPassword(createOpenAIKey());
+                        }
+                    };
+
+                    task.queue();
+                    currentKeyValue = task.getResult();
+                } else {
+                    currentKeyValue = PasswordSafe.getInstance().getPassword(createOpenAIKey());
+                }
+
+                cachedOpenAIKey = currentKeyValue;
+                // we're only caching successfully retrieved key values
+                isCached = true;
+            } catch (Exception e) {
+                isCached = false;
+                Logger.getInstance(this.getClass()).warn("Failed to fetch OpenAI key", e);
+                return null;
+            }
+        }
+
+        return cachedOpenAIKey;
     }
 
     @Override
     public void setOpenAIKey(@Nullable String key) {
         var oldValue = getOpenAIKey();
         try {
             PasswordSafe.getInstance().setPassword(createOpenAIKey(), key);
+
+            cachedOpenAIKey = key;
+            isCached = true;
```

Comment:
nit: `cachedOpenAIKey` is nullable, indicating whether or not a key has been cached. Maintaining the `isCached` flag isn't necessary.

Reply:
I also wanted to cache null values, i.e. the state when there's no stored OpenAI key. That's why I added the `isCached` flag.